### DISCUSSION TITLE
How long will `brew test` require HOMEBREW_NO_INSTALL_FROM_API?

### DISCUSSION THREAD

**Discussion by scpeters** [Homebrew/discussions#4237] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.1-60-g6ad9294-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6ad92949e910041416d84a53966ec46b873e069f
Last commit: 20 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 706bd6e2787e493263eebc031ec76b3f22f5cec4
Core tap last commit: 3 weeks ago
Core tap branch: master
Core tap JSON: 19 Feb 01:13 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.0SgO64EzRb/org.xquartz:0
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.39.0 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6.2-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
==> Downloading https://formulae.brew.sh/api/formula.json

Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  gz-sim7
  gazebo11

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> /Users/scpeters/.bash_profile
```

### Description of issue

We have some CI jobs for our 3rd-party tap (`osrf/simulation`) that install bottles and run `brew test $formula_name` to confirm that it currently works, but I see that `test` is currently one of the [commands that requires HOMEBREW_NO_INSTALL_FROM_API](https://github.com/Homebrew/brew/blob/6ad92949e910041416d84a53966ec46b873e069f/Library/Homebrew/brew.sh#L808). I haven't yet set `HOMEBREW_NO_INSTALL_FROM_API ` in these jobs and am seeing [failures](https://build.osrfoundation.org/job/ignition_math7-install_bottle-homebrew-amd64/113/console) of `brew test` in these CI jobs as the pre-existing local clone of `homebrew-core` on these machines is a bit out of date. I will go ahead and set `HOMEBREW_NO_INSTALL_FROM_API` on these jobs for now, but I'm just curious if there is a long-term plan for this variable. I assume we will keep using `git` as the backend of `homebrew-core` for the near future, so development activities will likely need `HOMEBREW_NO_INSTALL_FROM_API` to be set.

Things are a bit confusing at the moment that some tools are using `HOMEBREW_NO_INSTALL_FROM_API ` automatically and some aren't and it took me a while to figure out what is happening.

The following discussion is closest to what I was experiencing:

* https://github.com/orgs/Homebrew/discussions/4150

**Comment by apainintheneck** [comment]
I agree that ideally this wouldn't be necessary and we'd find some other way to handle the differences between the loading packages from the API and local taps in those commands where `HOMEBREW_NO_INSTALL_FROM_API` is set automatically. It's also causing problems with `brew audit` as outlined in https://github.com/Homebrew/brew/issues/14701.

I don't think there is any agreement though on what would generally be an improvement for that. Maybe this should be handled on a command-by-command basis.
